dbms/src/Storages/Page/V3/PageDirectory.h

```diff
@@ -82,48 +82,170 @@ class PageDirectorySnapshot : public DB::PageStorageSnapshot
+    {
+        if (container == nullptr)
+        {
+            return nullptr;
+        }
+        auto * new_container = new MultiVersionRefCount();
+        for (const auto & [ver, ref_count] : container->ref_map)
+        {
+            new_container->ref_map.emplace(ver, ref_count);
+        }
+        return new_container;
+    }
+
+    static void destroy(MultiVersionRefCount * container)
+    {
+        delete container;
+    }
+
+    static void decrLatestRefCountInSnap(MultiVersionRefCount * container, UInt64 snap_seq, Int64 deref_count)
+    {
+        assert(container != nullptr);
+        if (snap_seq == 0)
+        {
+            RUNTIME_CHECK(!container->ref_map.empty());
+            auto iter = container->ref_map.end();
+            iter--;
+            RUNTIME_CHECK(iter->second > deref_count);
+            iter->second = iter->second - deref_count;
+            container->ref_map.erase(container->ref_map.begin(), iter);
+        }
+        else
+        {
+            auto iter = MapUtils::findMutLess(container->ref_map, PageVersion(snap_seq + 1));
+            RUNTIME_CHECK(iter != container->ref_map.end());
+            RUNTIME_CHECK(iter->second > deref_count);
+            auto old_iter = iter;
+            for (; iter != container->ref_map.end(); ++iter)
+            {
+                iter->second = iter->second - deref_count;
+            }
+            // older version is just need for gc.
+            // So when we decr ref with a snap, we can be sure that older versions won't be need anymore
+            container->ref_map.erase(container->ref_map.begin(), old_iter);
+        }
+    }
+
+    [[nodiscard]] static MultiVersionRefCount * appendRefCount(MultiVersionRefCount * container, const PageVersion & ver, Int64 ref_count)
+    {
+        RUNTIME_CHECK(ref_count > 1);
+        if (container == nullptr)
+        {
+            container = new MultiVersionRefCount();
+        }
+        RUNTIME_CHECK(container->ref_map.empty() || container->ref_map.rbegin()->first < ver);
+        container->ref_map.emplace(ver, ref_count);
+        return container;
+    }
+
+    static Int64 getRefCountInSnap(MultiVersionRefCount * container, UInt64 snap_seq)
+    {
+        if (container == nullptr)
+        {
+            return 1;
+        }
+        RUNTIME_CHECK(!container->ref_map.empty());
+        auto iter = MapUtils::findLess(container->ref_map, PageVersion(snap_seq + 1));
+        if (iter == container->ref_map.end())
+        {
+            return 1;
+        }
+        else
+        {
+            return iter->second;
+        }
+    }
+
+    static Int64 getLatestRefCount(MultiVersionRefCount * container)
+    {
+        if (container == nullptr)
+        {
+            return 1;
+        }
+        RUNTIME_CHECK(!container->ref_map.empty());
+        return container->ref_map.rbegin()->second;
+    }
+
+#ifndef DBMS_PUBLIC_GTEST
+private:
+#endif
+    std::map<PageVersion, Int64> ref_map;
```

Comment:
using a `std::unique_ptr<std::map<PageVersion, Int64>>` maybe more exception safe

Reply:
It is just a member, why use unique_ptr can make it more exception safe? Or do you mean to be careful with the exception when creating raw pointer of `MultiVersionRefCount`?

Reply:
As all methods are static, we don't have a deconstructor for `MultiVersionRefCount `. We were possibly not calling `MultiVersionRefCount::destroy` when the exception happened. It will cause a memory leak.

Reply:
Avoid static method now.